File: 2020/02.kt
import java.io.File

private const val FILENAME = "2020/02.txt"
private val REGEX = Regex("""(\d+)-(\d+)\s(\w):\s(\w+)""")

// No input validation done as it is always correct

fun main() {
    partOne()
    partTwo()
}

private fun partOne() {
    val lines = File(FILENAME).readLines()

    var correctPasswords = 0

    lines.forEach { line ->
        val (min, max, letter, password) = REGEX.matchEntire(line)!!.destructured
        val range = min.toInt()..max.toInt()
        val count = password.count { it == letter.single() }

        if (count in range) {
            correctPasswords++
        }
    }

    println(correctPasswords)
}

private fun partTwo() {
    val lines = File(FILENAME).readLines()

    var correctPasswords = 0

    lines.forEach { line ->
        val (firstPosition, secondPosition, letter, password) = REGEX.matchEntire(line)!!.destructured

        // Subtract one from indexes as input file starts indexing at 1, but string indexes from 0
        val first: Boolean = password[firstPosition.toInt() - 1] == letter.single()
        val second: Boolean = password[secondPosition.toInt() - 1] == letter.single()

        if (first xor second) {
            correctPasswords++
        }
    }

    println(correctPasswords)
}

File: 2022/02/02.py
FILENAME = "2022/02/input.txt"


def part_one():
    score = 0

    with open(FILENAME, "r", encoding="UTF-8") as file:
        for line in file:
            opponent, me = line.split()

            if me == "X":
                score += 1
                me = "A"
            elif me == "Y":
                score += 2
                me = "B"
            else:
                score += 3
                me = "C"

            if (
                opponent == "A"
                and me == "B"
                or opponent == "B"
                and me == "C"
                or opponent == "C"
                and me == "A"
            ):
                score += 6
            elif opponent == me:
                score += 3

    print(score)


def part_two():
    score = 0

    with open(FILENAME, "r", encoding="UTF-8") as file:
        for line in file:
            opponent, me = line.split()

            if me == "X":
                me = "A"
            elif me == "Y":
                score += 3
                me = "B"
            else:
                score += 6
                me = "C"

            if (
                opponent == "C"
                and me == "C"
                or opponent == "A"
                and me == "B"
                or opponent == "B"
                and me == "A"
            ):
                score += 1
            elif (
                opponent == "C"
                and me == "A"
                or opponent == "A"
                and me == "C"
                or opponent == "B"
                and me == "B"
            ):
                score += 2
            elif (
                opponent == "A"
                and me == "A"
                or opponent == "B"
                and me == "C"
                or opponent == "C"
                and me == "B"
            ):
                score += 3

    print(score)


if __name__ == "__main__":
    part_one()
    part_two()
